src/libertem/udf/base.py

```diff
@@ -2388,25 +2390,28 @@ def run_for_dataset_sync(
         damage.set_shape_ds(dataset.shape, roi)
         damage.allocate()
         any_result = False
-        for part_results, task in result_iter:
-            any_result = True
-            with tracer.start_as_current_span("_apply_part_result -> UDF.merge"):
-                self._apply_part_result(
-                    udfs=self._udfs,
-                    damage=damage,
-                    part_results=part_results,
-                    task=task
-                )
-            if iterate:
+        try:
+            for part_results, task in result_iter:
+                any_result = True
+                with tracer.start_as_current_span("_apply_part_result -> UDF.merge"):
+                    self._apply_part_result(
+                        udfs=self._udfs,
+                        damage=damage,
+                        part_results=part_results,
+                        task=task
+                    )
+                if iterate:
+                    yield self._make_udf_result(
+                        udfs=self._udfs,
+                        damage=damage
+                    )
+            if not any_result or not iterate:
                 yield self._make_udf_result(
                     udfs=self._udfs,
                     damage=damage
                 )
-        if not any_result or not iterate:
-            yield self._make_udf_result(
-                udfs=self._udfs,
-                damage=damage
-            )
+        except JobCancelledError:
+            raise UDFRunCancelled
```

Comment:
Could consider a descriptive error message ? If we change `any_result` to an integer which we increment each time through to loop:

```python
        any_result = 0
        try:
            for part_results, task in result_iter:
                any_result += 1
                with tracer.start_as_current_span("_apply_part_result -> UDF.merge"):
                    self._apply_part_result(
                        udfs=self._udfs,
                        damage=damage,
                        part_results=part_results,
                        task=task
                    )
                if iterate:
                    yield self._make_udf_result(
                        udfs=self._udfs,
                        damage=damage
                    )
            if not any_result or not iterate:
                yield self._make_udf_result(
                    udfs=self._udfs,
                    damage=damage
                )
        except JobCancelledError:
            raise UDFRunCancelled(f'UDF run cancelled after {any_result} partitions')
```

Unfortunately we can't show the expected number of tasks because it's not exposed on the `result_iter`, I think.

Reply:
Good point, I'll improve the error message

Reply:
This is now implemented in 2ebac2bcf6ab42d82767f30c3adf3998aafc606e, thanks for the feedback!

Reply:
Great, looks good.